File: nim/src/dsa/data_structures/binary_search_tree.nim
## Binary Search Tree - Ordered binary tree
##
## Time Complexity (average):
## - insert: O(log n)
## - contains: O(log n)
## - remove: O(log n)
## - findMin/findMax: O(log n)

type
  BSTNode[T] = ref object
    data: T
    left: BSTNode[T]
    right: BSTNode[T]

  BinarySearchTree*[T] = object
    root: BSTNode[T]
    count: int

proc newBinarySearchTree*[T](): BinarySearchTree[T] =
  ## Create a new empty binary search tree
  BinarySearchTree[T](root: nil, count: 0)

proc insertNode[T](node: BSTNode[T], value: T): BSTNode[T] =
  if node == nil:
    return BSTNode[T](data: value, left: nil, right: nil)
  if value < node.data:
    node.left = insertNode(node.left, value)
  elif value > node.data:
    node.right = insertNode(node.right, value)
  node

proc insert*[T](tree: var BinarySearchTree[T], value: T) =
  ## Insert a value into the tree
  let oldRoot = tree.root
  tree.root = insertNode(tree.root, value)
  if tree.root != oldRoot or tree.count == 0:
    inc tree.count

proc containsNode[T](node: BSTNode[T], value: T): bool =
  if node == nil:
    return false
  if value == node.data:
    return true
  if value < node.data:
    return containsNode(node.left, value)
  containsNode(node.right, value)

proc contains*[T](tree: BinarySearchTree[T], value: T): bool =
  ## Check if value exists in the tree
  containsNode(tree.root, value)

proc findMinNode[T](node: BSTNode[T]): BSTNode[T] =
  var current = node
  while current.left != nil:
    current = current.left
  current

proc findMaxNode[T](node: BSTNode[T]): BSTNode[T] =
  var current = node
  while current.right != nil:
    current = current.right
  current

proc findMin*[T](tree: BinarySearchTree[T]): T =
  ## Find the minimum value
  ## Raises ValueError if tree is empty
  if tree.root == nil:
    raise newException(ValueError, "Tree is empty")
  findMinNode(tree.root).data

proc findMax*[T](tree: BinarySearchTree[T]): T =
  ## Find the maximum value
  ## Raises ValueError if tree is empty
  if tree.root == nil:
    raise newException(ValueError, "Tree is empty")
  findMaxNode(tree.root).data

proc removeNode[T](node: BSTNode[T], value: T): (BSTNode[T], bool) =
  if node == nil:
    return (nil, false)

  if value < node.data:
    let (newLeft, removed) = removeNode(node.left, value)
    node.left = newLeft
    return (node, removed)
  elif value > node.data:
    let (newRight, removed) = removeNode(node.right, value)
    node.right = newRight
    return (node, removed)
  else:
    # Found node to remove
    if node.left == nil:
      return (node.right, true)
    elif node.right == nil:
      return (node.left, true)
    else:
      # Node has two children
      let successor = findMinNode(node.right)
      node.data = successor.data
      let (newRight, _) = removeNode(node.right, successor.data)
      node.right = newRight
      return (node, true)

proc remove*[T](tree: var BinarySearchTree[T], value: T): bool =
  ## Remove a value from the tree
  ## Returns true if value was found and removed
  let (newRoot, removed) = removeNode(tree.root, value)
  tree.root = newRoot
  if removed:
    dec tree.count
  removed

proc isEmpty*[T](tree: BinarySearchTree[T]): bool =
  ## Check if the tree is empty
  tree.count == 0

proc size*[T](tree: BinarySearchTree[T]): int =
  ## Return the number of nodes
  tree.count

proc len*[T](tree: BinarySearchTree[T]): int =
  ## Return the number of nodes (alias for size)
  tree.count

proc clear*[T](tree: var BinarySearchTree[T]) =
  ## Remove all nodes
  tree.root = nil
  tree.count = 0

proc inorderNode[T](node: BSTNode[T], result: var seq[T]) =
  if node != nil:
    inorderNode(node.left, result)
    result.add(node.data)
    inorderNode(node.right, result)

proc inorder*[T](tree: BinarySearchTree[T]): seq[T] =
  ## Get values in sorted order (in-order traversal)
  result = @[]
  inorderNode(tree.root, result)

proc preorderNode[T](node: BSTNode[T], result: var seq[T]) =
  if node != nil:
    result.add(node.data)
    preorderNode(node.left, result)
    preorderNode(node.right, result)

proc preorder*[T](tree: BinarySearchTree[T]): seq[T] =
  ## Get values in pre-order traversal
  result = @[]
  preorderNode(tree.root, result)

proc postorderNode[T](node: BSTNode[T], result: var seq[T]) =
  if node != nil:
    postorderNode(node.left, result)
    postorderNode(node.right, result)
    result.add(node.data)

proc postorder*[T](tree: BinarySearchTree[T]): seq[T] =
  ## Get values in post-order traversal
  result = @[]
  postorderNode(tree.root, result)

proc heightNode[T](node: BSTNode[T]): int =
  if node == nil:
    return -1
  max(heightNode(node.left), heightNode(node.right)) + 1

proc height*[T](tree: BinarySearchTree[T]): int =
  ## Get the height of the tree
  heightNode(tree.root)

iterator items*[T](tree: BinarySearchTree[T]): T =
  ## Iterate over values in sorted order
  for value in tree.inorder():
    yield value

when isMainModule:
  var tree = newBinarySearchTree[int]()
  tree.insert(5)
  tree.insert(3)
  tree.insert(7)
  tree.insert(1)
  tree.insert(9)
  assert tree.contains(5)
  assert tree.contains(3)
  assert not tree.contains(6)
  assert tree.findMin() == 1
  assert tree.findMax() == 9
  assert tree.size() == 5
  assert tree.remove(3)
  assert not tree.contains(3)
  assert tree.size() == 4
  assert tree.inorder() == @[1, 5, 7, 9]
  echo "BinarySearchTree tests passed!"

File: python/src/dsa/data_structures/binary_search_tree.py
"""Binary search tree implementation.

Time: O(log n) average, O(n) worst for insert, search, delete
Space: O(n)
"""

from typing import Generic, TypeVar, Optional, Callable

T = TypeVar("T")


class _Node(Generic[T]):
    """Internal node class."""

    def __init__(self, value: T) -> None:
        self.value = value
        self.left: Optional[_Node[T]] = None
        self.right: Optional[_Node[T]] = None


class BinarySearchTree(Generic[T]):
    """A generic binary search tree implementation."""

    def __init__(self, key: Optional[Callable[[T], any]] = None) -> None:
        self._root: Optional[_Node[T]] = None
        self._len = 0
        self._key = key or (lambda x: x)

    def insert(self, value: T) -> bool:
        """Insert a value. Returns True if inserted, False if duplicate."""
        if self._root is None:
            self._root = _Node(value)
            self._len += 1
            return True
        return self._insert(self._root, value)

    def _insert(self, node: _Node[T], value: T) -> bool:
        key_val = self._key(value)
        node_key = self._key(node.value)

        if key_val < node_key:
            if node.left is None:
                node.left = _Node(value)
                self._len += 1
                return True
            return self._insert(node.left, value)
        elif key_val > node_key:
            if node.right is None:
                node.right = _Node(value)
                self._len += 1
                return True
            return self._insert(node.right, value)
        return False  # Duplicate

    def contains(self, value: T) -> bool:
        """Return True if the value is in the tree."""
        return self._search(self._root, value) is not None

    def _search(self, node: Optional[_Node[T]], value: T) -> Optional[_Node[T]]:
        if node is None:
            return None
        key_val = self._key(value)
        node_key = self._key(node.value)

        if key_val < node_key:
            return self._search(node.left, value)
        elif key_val > node_key:
            return self._search(node.right, value)
        return node

    def remove(self, value: T) -> bool:
        """Remove a value. Returns True if removed, False if not found."""
        self._root, removed = self._remove(self._root, value)
        if removed:
            self._len -= 1
        return removed

    def _remove(
        self, node: Optional[_Node[T]], value: T
    ) -> tuple[Optional[_Node[T]], bool]:
        if node is None:
            return None, False

        key_val = self._key(value)
        node_key = self._key(node.value)

        if key_val < node_key:
            node.left, removed = self._remove(node.left, value)
            return node, removed
        elif key_val > node_key:
            node.right, removed = self._remove(node.right, value)
            return node, removed

        # Found the node to remove
        if node.left is None:
            return node.right, True
        if node.right is None:
            return node.left, True

        # Two children: find minimum in right subtree
        min_node = self._find_min(node.right)
        node.value = min_node.value
        node.right, _ = self._remove(node.right, min_node.value)
        return node, True

    def _find_min(self, node: _Node[T]) -> _Node[T]:
        while node.left is not None:
            node = node.left
        return node

    def min(self) -> Optional[T]:
        """Return the minimum value, or None if empty."""
        if self._root is None:
            return None
        return self._find_min(self._root).value

    def max(self) -> Optional[T]:
        """Return the maximum value, or None if empty."""
        if self._root is None:
            return None
        node = self._root
        while node.right is not None:
            node = node.right
        return node.value

    def inorder(self) -> list[T]:
        """Return values in sorted order."""
        result: list[T] = []
        self._inorder(self._root, result)
        return result

    def _inorder(self, node: Optional[_Node[T]], result: list[T]) -> None:
        if node is not None:
            self._inorder(node.left, result)
            result.append(node.value)
            self._inorder(node.right, result)

    def is_empty(self) -> bool:
        """Return True if the tree is empty."""
        return self._root is None

    def __len__(self) -> int:
        """Return the number of elements."""
        return self._len

    def __contains__(self, value: T) -> bool:
        """Check if value exists."""
        return self.contains(value)
